warehouse/views.py

```diff
@@ -250,20 +251,36 @@ def index(request):
 def search(request):
 
     q = request.params.get("q", '')
+    q = q.replace("'", '"')
+
+    def filter_query(s):
+        matches = re.findall(r'(?:"([^"]*)")|([^"]*)', s)
+        result_quoted = [t[0].strip() for t in matches if t[0]]
+        result_unquoted = [t[1].strip() for t in matches if t[1]]
+        return result_quoted, result_unquoted
+
+    def form_query(query_type, query):
+        return Q('multi_match', fields=SEARCH_FIELDS,
+                 query=query, type=query_type
+                 )
 
     if q:
-        should = []
-        for field in SEARCH_FIELDS:
-            kw = {"query": q}
-            if field in SEARCH_BOOSTS:
-                kw["boost"] = SEARCH_BOOSTS[field]
-            should.append(Q("match", **{field: kw}))
-
-        # Add a prefix query if ``q`` is longer than one character.
+        quoted_string, unquoted_string = filter_query(q)
+        must = [
+            form_query("phrase", i) for i in quoted_string
+        ] + [
+            form_query("best_fields", i) for i in unquoted_string
+        ]
+
+        bool_query = Q('bool', must=must)
+
+        # Allow to optionally match on prefix
+        # if ``q`` is longer than one character.
         if len(q) > 1:
-            should.append(Q('prefix', normalized_name=q))
+            bool_query = bool_query | Q('prefix', normalized_name=q)
+
+        query = request.es.query('bool', must=bool_query)
```

Comment:
there is no need to wrap the query in a `bool`. It doesn't do anything.

Reply:
@HonzaKral  The reason why I did it is because in the next line we're using the `suggest` method on query and on the `Q` object i.e `bool_query` doesn't have a suggest method and would throw an error.

Reply:
yes, you need to use ` Search` object, but that accepts any query, so just `request.es.query(query)` should do the trick.